File: python/core/registration.py
import time

from .config import get_config
from .mq import publish_json_message


def publish_registration(
    *,
    event_type: str,
    instance_id: str,
    tags: list[str] | None = None,
) -> None:
    cfg = get_config()

    publish_json_message(
        host=cfg.rabbitmq_host,
        port=cfg.rabbitmq_port,
        exchange=cfg.registry_exchange,
        exchange_type='topic',
        routing_key=event_type,
        message={
            'event_type': event_type,
            'instance_id': instance_id,
            'tags': tags or [],
        },
        username=cfg.rabbitmq_username,
        password=cfg.rabbitmq_password,
    )


def heartbeat_loop(
    *,
    instance_id: str,
    tags: list[str] | None = None,
    interval_seconds: int = 60,
) -> None:
    while True:
        publish_registration(
            event_type='connector.heartbeat',
            instance_id=instance_id,
            tags=tags,
        )
        time.sleep(interval_seconds)

File: typescript/src/core/registration.ts
import { getConfig } from './config.js';
import { publishJsonMessage } from './mqPublisher.js';

export async function publishRegistration(opts: {
  eventType: string;
  instanceId: string;
  tags?: string[];
}): Promise<void> {
  const cfg = getConfig();

  await publishJsonMessage({
    cfg,
    exchange: cfg.registryExchange,
    exchangeType: 'topic',
    routingKey: opts.eventType,
    message: {
      event_type: opts.eventType,
      instance_id: opts.instanceId,
      tags: opts.tags ?? [],
    },
  });
}

export function startHeartbeatLoop(opts: {
  instanceId: string;
  tags?: string[];
  intervalSeconds: number;
}): void {
  const tick = async () => {
    try {
      await publishRegistration({
        eventType: 'connector.heartbeat',
        instanceId: opts.instanceId,
        tags: opts.tags,
      });
    } catch {
      /* heartbeat failures are non-fatal */
    }
  };

  setInterval(tick, opts.intervalSeconds * 1000);
}
